packages/protocol/contracts/ACE/ACE.sol

```diff
@@ -133,138 +180,336 @@ contract ACE {
-        );
-        noteRegistries[msg.sender] = registry;
-        return address(registry);
+        bool _canConvert
+    ) public {
+        require(registries[msg.sender].flags.active == false, "address already has a linked note registry");
+        NoteRegistry memory registry = NoteRegistry({
+            linkedToken: ERC20(_linkedTokenAddress),
+            scalingFactor: _scalingFactor,
+            totalSupply: 0,
+            confidentialTotalSupply: bytes32(0x0),
+            flags: Flags({
+                active: true,
+                canMint: _canMint,
+                canBurn: _canBurn,
+                canConvert: _canConvert
+            })
+        });
+        registries[msg.sender] = registry;
     }
 
     function updateNoteRegistry(
-        bytes memory _proofOutput, 
-        uint16 _proofType, 
-        address _proofSender
-    ) public returns (bool) {
-        NoteRegistry registry = noteRegistries[msg.sender];
-        require(registry != NoteRegistry(0), "sender does not have a linked Note Registry");
-        require(registry.updateNoteRegistry(_proofOutput, _proofType, _proofSender), "update failed!");
-        return true;
+        uint24 _proof,
+        address _proofSender,
+        bytes memory _proofOutput
+    ) public {
+        NoteRegistry storage registry = registries[msg.sender];
+        require(registry.flags.active == true, "note registry does not exist for the given address");
+        bytes32 proofHash = keccak256(_proofOutput);
+        require(
+            validateProofByHash(_proof, proofHash, _proofSender) == true,
+            "ACE has not validated a matching proof"
+        );
+
+        (bytes memory inputNotes,
+        bytes memory outputNotes,
+        address publicOwner,
+        int256 publicValue) = _proofOutput.extractProofOutput();
+
+        updateInputNotes(inputNotes);
+        updateOutputNotes(outputNotes);
+
+        if (publicValue != 0) {
+            require(registry.flags.canMint == false, "mintable assets cannot be converted into public tokens");
+            require(registry.flags.canBurn == false, "burnable assets cannot be converted into public tokens");
+            require(registry.flags.canConvert == true, "this asset cannot be converted into public tokens");
+            if (publicValue < 0) {
+                registry.totalSupply = registry.totalSupply.add(uint256(-publicValue));
+                require(
+                    registry.publicApprovals[publicOwner][proofHash] >= uint256(-publicValue),
+                    "public owner has not validated a transfer of tokens"
+                );
+                registry.publicApprovals[publicOwner][proofHash] -= uint256(-publicValue);
+                require(
+                    registry.linkedToken.transferFrom(publicOwner, address(this), uint256(-publicValue)), 
+                    "transfer failed"
+                );
+            } else {
+                registry.totalSupply -= uint256(publicValue);
```

Comment:
SafeMath.sub

Reply:
Fixed in https://github.com/AztecProtocol/AZTEC/pull/75